service/history/workflow/context.go

```diff
@@ -698,48 +697,67 @@ func (c *ContextImpl) SubmitClosedWorkflowSnapshot(
 	)
 }
 
-func (c *ContextImpl) mergeContinueAsNewReplicationTasks(
-	updateMode persistence.UpdateWorkflowMode,
+func (c *ContextImpl) mergeUpdateWithNewReplicationTasks(
 	currentWorkflowMutation *persistence.WorkflowMutation,
 	newWorkflowSnapshot *persistence.WorkflowSnapshot,
 ) error {
 
-	if currentWorkflowMutation.ExecutionState.Status != enumspb.WORKFLOW_EXECUTION_STATUS_CONTINUED_AS_NEW {
+	if newWorkflowSnapshot == nil {
 		return nil
-	} else if updateMode == persistence.UpdateWorkflowModeBypassCurrent && newWorkflowSnapshot == nil {
-		// update current workflow as zombie & continue as new without new zombie workflow
-		// this case can be valid if new workflow is already created by resend
+	}
+
+	// TODO: looks like ndc tests will generate this invalid case
+	// if currentWorkflowMutation.ExecutionState.Status == enumspb.WORKFLOW_EXECUTION_STATUS_RUNNING {
+	// 	return serviceerror.NewInternal("cannot creating new workflow when current workflow is still running")
+	// }
+
+	if currentWorkflowMutation.ExecutionState.Status != enumspb.WORKFLOW_EXECUTION_STATUS_CONTINUED_AS_NEW &&
+		!c.config.ReplicationEnableUpdateWithNewTaskMerge() {
+		// we need to make sure target cell is able to handle updateWithNew
+		// for non continuedAsNewcase before enabling this feature
 		return nil
 	}
 
-	// current workflow is doing continue as new
+	// current workflow is doing closing with new workflow
 
+	// namespace could be local or global with only one cluster, so no replication task is generated.
+	// TODO: What does the following comment mean?
 	// it is possible that continue as new is done as part of passive logic
-	if len(currentWorkflowMutation.Tasks[tasks.CategoryReplication]) == 0 {
+	numCurrentReplicationTasks := len(currentWorkflowMutation.Tasks[tasks.CategoryReplication])
+	numNewReplicationTasks := len(newWorkflowSnapshot.Tasks[tasks.CategoryReplication])
+
+	if numCurrentReplicationTasks == 0 && numNewReplicationTasks == 0 {
 		return nil
 	}
-
-	if newWorkflowSnapshot == nil || len(newWorkflowSnapshot.Tasks[tasks.CategoryReplication]) != 1 {
-		return serviceerror.NewInternal("unable to find replication task from new workflow for continue as new replication")
+	if numCurrentReplicationTasks == 0 {
+		return serviceerror.NewInternal("current workflow has no replication task, while new workflow contains replication task")
+	}
+	if numNewReplicationTasks != 1 {
+		// TODO: support more than one replication tasks (batch of events) in the new workflow
```

Comment:
I am not sure what is the use case to support more than one replication task

Reply:
Not necessary workflow. I am thinking in general when starting a state machine it can perform multiple transitions (result in multiple history batches) in one transaction, for example the new multi-operation API we are adding. 
But this also depends on the implementation of replication task generation. We can change that to only generate one replication task for one transaction regardless of the # of transitions or history batches, then the logic here doesn't need to be changed.